### PR TITLE
create module heart

### PR DESCRIPTION
**Pull request by poptart-coral**

### CODE REVIEW

**Comment by poptart-coral** [comment]
probleme de format de date, avec vine je peux utiliser que date, pas datetime, ce qui faisait une incompatibilité avec le type DateTime dans le model Heart, j'ai donc tout passé en Date mais je ne suis pas sûre qu'il faille faire ça

File: apps/heart/validators/heart.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  vine.object({
+    patientId: vine.string(),
+    startDate: vine.string(),
+    endDate: vine.string(),
+    minHeartRate: vine.number(),
+    maxHeartRate: vine.number(),
+  })
+)
+
+export const HeartQueryValidator = vine.compile(
```
Comment (NathaelB) [inline apps/heart/validators/heart.ts]: You can rename it to `heartQueryValidator` and create a interface `HeartQuerySchema`

File: apps/heart/controllers/patient_heart_rates_controller.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+@inject()
+export default class PatientHeartRatesController {
+  constructor(private heartService: PatientHeartRatesService) {}
+
+  async show({ params }: HttpContext) {
+    logger.info('Fetching heart rate for patient %s', params.id)
+    const data = await params.validateUsing(HeartQueryValidator)
+
+    return this.heartService.findByPatientId(data.patientId)
```
Comment (NathaelB) [inline apps/heart/controllers/patient_heart_rates_controller.ts]: You've defined the `patientId` as optional, so if it's undefined you'll have a problem.
In the controller behaviour, if the patientId is defined then it's often a professional who will check the data. If no patientId is provided, we use the identifier of the connected patient.

File: apps/heart/services/patient_heart_rates_service.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import Heart from '#models/heart'
+import { CreateHeartSchema } from '#apps/heart/validators/heart'
+import logger from '@adonisjs/core/services/logger'
+import HeartException from '#apps/heart/exceptions/heart_exception'
+
+export default class PatientHeartRatesService {
+  async findByPatientId(patientId: string) {
```
Comment (NathaelB) [inline apps/heart/services/patient_heart_rates_service.ts]: Your validator takes several parameters, but they're not used here, is that normal?

**Review by NathaelB — CHANGES_REQUESTED** [review]
The code as a whole is clean, you've separated the behaviours in the controller and in the service. When creating a resource, you manage errors by raising exceptions and handling them correctly, which is great.

I've commented on certain parts to take into account the parameters passed to a request.

Once the authorisation system has been established, we can add this part later!

Excellent work!

File: apps/heart/controllers/patient_heart_rates_controller.ts (hunk 2 above)
Comment (poptart-coral) [reply apps/heart/controllers/patient_heart_rates_controller.ts]: "we use the identifier of the connected patient." how should i get it ?

File: apps/heart/services/patient_heart_rates_service.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      })
+  }
+
+  async create(payload: CreateHeartSchema): Promise<Heart> {
+    try {
+      return Heart.create(payload)
+    } catch (err) {
+      logger.error(err)
+      throw new HeartException('Failed to create heart rate data', {
+        code: 'E_HEART_CREATE',
```
Comment (Courtcircuits) [inline apps/heart/services/patient_heart_rates_service.ts]: Rename this into `E_HEART_RATE_CREATE`

Comment (poptart-coral) [reply apps/heart/services/patient_heart_rates_service.ts]: yes yes

File: apps/shared/models/heart.ts
Comment (NathaelB) [inline apps/shared/models/heart.ts:1]: In the validator to manage a patient's heart rate, it was specified to have an array like this:

```json
{
  "patientId": "123456",
  "heartRates": [
    {
      "startDate": "2024-06-30T19:42:00Z",
      "endDate": "2024-06-30T19:42:00Z",
      "heartRate": 100,
    },
  ]
}
```

Looking around, here you've created a `heart` model that represents heart rate? If so, there are a few problems:
- The name is too general, `heart_rate` would be more appropriate
- Why model ranges for the time and value metrics?

File: apps/heart/validators/heart.ts
Comment (NathaelB) [inline apps/heart/validators/heart.ts:1]: Heart represents the entire module, as a kind of service.

Here in the issue we were asked to manage the heart rate part. So heart_rate would be more appropriate.
Note that the `createHeartValidator` does not respect the contract established in the issue.

image

File: apps/heart/controllers/patient_heart_rates_controller.ts
Comment (NathaelB) [inline apps/heart/controllers/patient_heart_rates_controller.ts:1]: I've modified the controller to show you an example of API documentation.
I've also changed the name of the service: `heartService` -> `patientHeartRatesService`

**Review by NathaelB — CHANGES_REQUESTED** [review]

**Review by NathaelB — APPROVED** [review]

**Pull request merged**